test/common/router/config_impl_test.cc

```diff
@@ -4201,6 +4207,163 @@ name: foo
               config.route(headers, 0)->routeEntry()->clusterName());
   }
 }
+
+TEST(PerFilterConfigsTest, PerFilterConfigs) {
+  Server::Configuration::BufferFilterConfig factory;
+  Registry::InjectFactory<Server::Configuration::NamedHttpFilterConfigFactory> register_buffer(
+      factory);
+
+  std::string yaml = R"EOF(
+max_request_bytes: 123
+max_request_time:
+  seconds: 456
+  nanos: 789
+)EOF";
+
+  ProtobufWkt::Struct buffer_cfg;
+  MessageUtil::loadFromYaml(yaml, buffer_cfg);
+  Protobuf::Map<std::string, ProtobufWkt::Struct> proto_cfgs;
+  proto_cfgs[factory.name()] = buffer_cfg;
+  PerFilterConfigs configs{proto_cfgs};
+
+  ASSERT_EQ(nullptr, configs.get("unknown.filter"))
+      << "filter configs that aren't present should return nullptr";
+
+  const Protobuf::Message* processed = configs.get(factory.name());
+  ASSERT_NE(nullptr, processed) << "filter configs present should return a concrete Message";
+
+  ASSERT_NO_THROW({
+    const envoy::config::filter::http::buffer::v2::Buffer* buf =
+        dynamic_cast<const envoy::config::filter::http::buffer::v2::Buffer*>(processed);
+
+    ASSERT_EQ(123, buf->max_request_bytes().value());
+    ASSERT_EQ(456, buf->max_request_time().seconds());
+    ASSERT_EQ(789, buf->max_request_time().nanos());
+  }) << "returned message should be castable to known type";
+}
+
+TEST(PerFilterConfigsTest, UnknownFilter) {
+  std::string yaml = R"EOF(
+foo: "bar"
+)EOF";
+
+  ProtobufWkt::Struct some_cfg;
+  MessageUtil::loadFromYaml(yaml, some_cfg);
+  Protobuf::Map<std::string, ProtobufWkt::Struct> proto_cfgs;
+  proto_cfgs["unknown.filter"] = some_cfg;
+
+  ASSERT_THROW(PerFilterConfigs{proto_cfgs}, EnvoyException)
+      << "should throw on unrecognized filter";
+}
+
+void checkPerFilterConfig(const Protobuf::Message* cfg, uint32_t expected_max_bytes,
+                          std::string source) {
+  ASSERT_NE(nullptr, cfg) << "config should not be null for source: " << source;
+  ASSERT_NO_THROW({
+    const auto buffer = dynamic_cast<const envoy::config::filter::http::buffer::v2::Buffer*>(cfg);
+    ASSERT_EQ(expected_max_bytes, buffer->max_request_bytes().value())
+        << "config value does not match expected for source: " << source;
+  }) << "config should properly dynamic_cast to the appropriate type for source: "
+     << source;
+}
+
+TEST(PerFilterConfigsTest, RouteLocalConfig) {
+  Server::Configuration::BufferFilterConfig factory;
```

Comment:
Instead of using buffer filter, can you register a test filter for this test, so that you can correctly verify that you can override to be a different type for the filter data? If you need to you can have a proto just for testing.

Reply:
is there an existing one I can use for this or should I create one? Should I create test proto in envoy or data plane?

Reply:
Just make one local to the test.